scripts/sensor/03b-time_frequency_csp.py

```diff
@@ -0,0 +1,988 @@
+            processing='csp+freq',
+            subject=subject,
+            session=session,
+            suffix='scores+std',
+            extension='.npy')
+
+    def tf_scores(self, subject: str, session: Session) -> BIDSPath:
+        """Path to time-frequency scores."""
+        return self.bids_basename.copy().update(
+            processing='csp+tf',
+            subject=subject,
+            session=session,
+            suffix='scores+std',
+            extension='.npy')
+
+    def prefix(self, subject: str, session: Session) -> str:
+        """Usefull when printing messages."""
+        if session:
+            return f'subject-{subject}-session-{session}'
+        else:
+            return f'subject-{subject}'
+
+
+class Tf:
+    """Util class containing useful info about the time frequency windows."""
+
+    def __init__(self, cfg):
+        """Calculate the required time and frequency size."""
+        freqs = np.array(cfg.csp_freqs)
+        times = np.array(cfg.csp_times)
+
+        freq_ranges = list(zip(freqs[:-1], freqs[1:]))
+        time_ranges = list(zip(times[:-1], times[1:]))
+
+        n_freq_windows = len(freq_ranges)
+        n_time_windows = len(time_ranges)
+
+        # For band passed periodic signal,
+        # according to the Nyquist theorem,
+        # we can reconstruct the signal if f_s > 2 * band_freq
+        min_band_freq = np.min(freqs[1:] - freqs[:-1])
+        min_band_time = np.min(times[1:] - times[:-1])
+        recommanded_w_min_time = 1 / (2 * min_band_freq)
+
+        if recommanded_w_min_time > min_band_time:
+            msg = ("We recommand increasing the duration of "
+                   "your time intervals "
+                   f"to at least {round(recommanded_w_min_time, 2)}s.")
+            logger.info(gen_log_message(msg, step=8))
+
+        centered_w_times = (times[1:] + times[:-1])/2
+        centered_w_freqs = (freqs[1:] + freqs[:-1])/2
+
+        self.freqs = freqs
+        self.freq_ranges = freq_ranges
+        self.times = times
+        self.time_ranges = time_ranges
+        self.centered_w_times = centered_w_times
+        self.centered_w_freqs = centered_w_freqs
+        self.n_time_windows = n_time_windows
+        self.n_freq_windows = n_freq_windows
+
+
+def prepare_labels(*, epochs: BaseEpochs, cfg) -> np.ndarray:
+    """Return the projection of the events_id on a boolean vector.
+
+    This projection is useful in the case of hierarchical events:
+    we project the different events contained in one condition into
+    just one label.
+
+    Returns:
+    --------
+    A boolean numpy array containing the labels.
+    """
+    tf_cond = cfg.time_frequency_conditions
+    epochs_cond_0 = epochs[tf_cond[0]]
+    event_id_condition_0 = set(epochs_cond_0.events[:, 2])
+    epochs_cond_1 = epochs[tf_cond[1]]
+    event_id_condition_1 = set(epochs_cond_1.events[:, 2])
+
+    y = epochs.events[:, 2].copy()
+    for i in range(len(y)):
```

Comment:
Style: use `enumerate()` in situations like these

Reply:
But here I think we cannot spare the usage of brackets. Even with enumerate, we still would have to use y[i], no ?

Reply:
Because we wave to modify the value, not just read it